WordPressCom-Stats-iOS/WPStatsService.m

```diff
@@ -56,26 +85,87 @@ - (void)retrieveAllStatsForDates:(NSArray *)dates
         return;
     }
     
-    NSMutableArray *endDates = [NSMutableArray new];
-    for (NSDate *date in dates) {
-        NSDate *endDate = [self calculateEndDateForPeriodUnit:unit withDateWithinPeriod:date];
-        [endDates addObject:endDate];
-    }
-
-    [self.remote batchFetchStatsForDates:endDates
-                                 andUnit:unit
-             withVisitsCompletionHandler:^(StatsVisits *visits, NSError *error)
-    {
+    NSDate *endDate = [self calculateEndDateForPeriodUnit:unit withDateWithinPeriod:date];
+    NSMutableDictionary *cacheDictionary = [self.ephemory objectForKey:@[@(unit), endDate]];
+    if (cacheDictionary) {
         if (visitsCompletion) {
-            visitsCompletion(visits, error);
+            visitsCompletion(cacheDictionary[@(StatsCacheVisits)], nil);
+        }
+
+        if (eventsCompletion) {
+            eventsCompletion(cacheDictionary[@(StatsCacheEvents)], nil);
+        }
+
+        if (postsCompletion) {
+            postsCompletion(cacheDictionary[@(StatsCachePosts)], nil);
+        }
+    
+        if (referrersCompletion) {
+            referrersCompletion(cacheDictionary[@(StatsCacheReferrers)], nil);
+        }
+    
+        if (clicksCompletion) {
+            clicksCompletion(cacheDictionary[@(StatsCacheClicks)], nil);
+        }
+    
+        if (countryCompletion) {
+            countryCompletion(cacheDictionary[@(StatsCacheCountry)], nil);
+        }
+    
+        if (videosCompletion) {
+            videosCompletion(cacheDictionary[@(StatsCacheVideos)], nil);
+        }
+    
+        if (commentsAuthorsCompletion) {
+            commentsAuthorsCompletion(cacheDictionary[@(StatsCacheCommentsAuthors)], nil);
+        }
+        
+        if (commentsPostsCompletion) {
+            commentsPostsCompletion(cacheDictionary[@(StatsCacheCommentsPosts)], nil);
+        }
+    
+        if (tagsCategoriesCompletion) {
+            tagsCategoriesCompletion(cacheDictionary[@(StatsCacheTagsCategories)], nil);
         }
+    
+        if (followersDotComCompletion) {
+            followersDotComCompletion(cacheDictionary[@(StatsCacheFollowersDotCom)], nil);
+        }
+    
+        if (followersEmailCompletion) {
+            followersEmailCompletion(cacheDictionary[@(StatsCacheFollowersEmail)], nil);
+        }
+    
+        if (publicizeCompletion) {
+            publicizeCompletion(cacheDictionary[@(StatsCachePublicize)], nil);
+        }
+        
+        completionHandler();
+        
+        return;
+    } else {
+        cacheDictionary = [NSMutableDictionary new];
+        [self.ephemory setObject:cacheDictionary forKey:@[@(unit), endDate]];
     }
+
+    [self.remote batchFetchStatsForDate:endDate
+                                andUnit:unit
+            withVisitsCompletionHandler:^(StatsVisits *visits, NSError *error)
+     {
+         cacheDictionary[@(StatsCacheVisits)] = visits;
```

Comment:
if you get an error on the remote call, will the visits be nil and then the cache insertion fail?
You may need to check for error first before doing this.

This happens for all completion handlers but I will mention only this time.